Lib/test/test_inspect/test_inspect.py

```diff
@@ -4027,44 +4027,217 @@ def __init__(self, b):
                            ('bar', 2, ..., "keyword_only")),
                           ...))
 
-    def test_signature_on_class_with_decorated_new(self):
+    def test_signature_on_class_with_decorated_init(self):
         def identity(func):
             @functools.wraps(func)
             def wrapped(*args, **kwargs):
                 return func(*args, **kwargs)
             return wrapped
 
-        class Foo:
+        class C:
             @identity
-            def __new__(cls, a, b):
+            def __init__(self, b):
                 pass
 
-        self.assertEqual(self.signature(Foo),
-                         ((('a', ..., ..., "positional_or_keyword"),
-                           ('b', ..., ..., "positional_or_keyword")),
-                          ...))
+        C(1)  # does not raise
+        self.assertEqual(self.signature(C),
+                        ((('b', ..., ..., "positional_or_keyword"),),
+                        ...))
 
-        self.assertEqual(self.signature(Foo.__new__),
-                         ((('cls', ..., ..., "positional_or_keyword"),
-                           ('a', ..., ..., "positional_or_keyword"),
-                           ('b', ..., ..., "positional_or_keyword")),
-                          ...))
+        with self.subTest('classmethod'):
+            class C:
+                @classmethod
+                @identity
+                def __init__(cls, b):
+                    pass
 
-        class Bar:
-            __new__ = identity(object.__new__)
+            C(1)  # does not raise
+            self.assertEqual(self.signature(C),
+                            ((('b', ..., ..., "positional_or_keyword"),),
+                            ...))
 
-        varargs_signature = (
-            (('args', ..., ..., 'var_positional'),
-             ('kwargs', ..., ..., 'var_keyword')),
-            ...,
-        )
+        with self.subTest('staticmethod'):
+            class C:
+                @staticmethod
+                @identity
+                def __init__(b):
+                    pass
+
+            C(1)  # does not raise
+            self.assertEqual(self.signature(C),
+                            ((('b', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('MethodType'):
+            class A:
+                @identity
+                def call(self, a):
+                    pass
+
+            class C:
+                __init__ = A().call
+
+            C(1)  # does not raise
+            self.assertEqual(self.signature(C),
+                            ((('a', ..., ..., "positional_or_keyword"),),
+                            ...))
+
+        with self.subTest('partial'):
+            class C:
+                __init__ = identity(functools.partial(lambda x, a, b: None, 2))
```

Comment:
Should not this be `functools.partial(identity(lambda x, a, b: None), 2)`? Although the "partial" and "partialmethod" subtests are not affected by all these changes, so I am not sure that it makes sense to keep them. If they do not fail, we cannot be sure that they are meaningful.

Reply:
Thank you for pointing out the error, I will correct this.

These test cases are based on the ones below (`test_signature_on_class_with_init` and `test_signature_on_class_with_new`), and I believe they need to be this detailed to catch potential bugs from changes made in `inspect.signature`'s behavior.

This test cases also fail in current version of cpython.